File: packages/lhc/src/messages/internal/project.ts
// Event to message + typed blocks. Verbatim means payload fields are copied
// into block content untouched: nothing here trims, normalizes, splits, or
// summarizes. Token estimates come from the one counting util, called directly:
// it is pure and deterministic, so golden counts beat stubs.
import { type ApiBlock, blobTokenEstimate, placeholderText } from "../../shared-tech/index.js";
import type { TokenEstimator } from "../../shared-tech/token-counting/index.js";
import type { Block, BlockType, RecordedEvent } from "../index.js";

// A message that carried content blocks beyond text keeps block 0 as its
// text-shaped form — the text of its text blocks with a short placeholder
// (type, media type, size, title) where each non-text block sits — so every
// reader of block 0 (bands, derivations, retrieval, token pricing) sees that
// the block existed without seeing its bytes. Rows 1..n hold the API blocks
// verbatim (blob payloads already replaced by references at intake), in order,
// so serving can replay the exact content array.
function apiBlockRows(blocks: readonly ApiBlock[] | undefined): Block[] {
  return (blocks ?? []).map((block) => ({ blockType: block.type as BlockType, content: block }));
}

function textShaped(text: string, blocks: readonly ApiBlock[] | undefined): string {
  if (blocks === undefined || blocks.length === 0) return text;
  return blocks
    .map(placeholderText)
    .filter((line) => line !== "")
    .join("\n");
}

function blobTokens(blocks: readonly ApiBlock[] | undefined): number {
  return (blocks ?? []).reduce((sum, block) => sum + blobTokenEstimate(block), 0);
}

export interface ProjectedMessage {
  blocks: Block[];
  // o200k text (+ blobs). Signature billed tokens are stored alongside this
  // number in token_estimate (see projectEvent) so durable rows stay one
  // integer; weighStored() subtracts the billed signature before weighting.
  tokenEstimate: number;
  signatureTokenEstimate: number;
}

function textRaw(estimator: TokenEstimator, text: string): number {
  return estimator.rawCount(text);
}

// turn_end is recorded in the event order but produces no message.
export function projectEvent(event: RecordedEvent, estimator: TokenEstimator): ProjectedMessage | null {
  switch (event.eventKind) {
    case "user_prompt": {
      const text = textShaped(event.payload.text, event.payload.blocks);
      return {
        blocks: [
          {
            blockType: "text",
            // The steer assertion rides the block so the record says why this
            // prompt sits mid-turn; text is what every reader consumes.
            content: event.payload.steer === true ? { text, steer: true } : { text },
          },
          ...apiBlockRows(event.payload.blocks),
        ],
        tokenEstimate: textRaw(estimator, text) + blobTokens(event.payload.blocks),
        signatureTokenEstimate: 0,
      };
    }
    case "runtime_note":
      return {
        blocks: [{ blockType: "text", content: { text: event.payload.text } }],
        tokenEstimate: textRaw(estimator, event.payload.text),
        signatureTokenEstimate: 0,
      };
    case "assistant_text": {
      const content: Record<string, unknown> = { text: event.payload.text };
      if (event.payload.provider !== undefined) content.provider = event.payload.provider;
      if (event.payload.model !== undefined) content.model = event.payload.model;
      if (event.payload.api !== undefined) content.api = event.payload.api;
      return {
        blocks: [{ blockType: "text", content }],
        tokenEstimate: textRaw(estimator, event.payload.text),
        signatureTokenEstimate: 0,
      };
    }
    case "assistant_thinking": {
      // Verbatim payload copy: text always; signature + model identity when sent.
      const content: Record<string, unknown> = { text: event.payload.text };
      if (event.payload.signature !== undefined) content.signature = event.payload.signature;
      if (event.payload.provider !== undefined) content.provider = event.payload.provider;
      if (event.payload.model !== undefined) content.model = event.payload.model;
      if (event.payload.api !== undefined) content.api = event.payload.api;
      // Replayed signatures are billed in the provider's input context and
      // must be counted, but at the measured provider rate rather than as BPE text.
      // token_estimate stores o200k(text) + billed signature so one integer stays
      // the durable size; signatureTokenEstimate lets weighStored skip re-weighting.
      const signatureTokens =
        event.payload.signature === undefined ? 0 : estimator.estimateSignature(event.payload.signature);
      return {
        blocks: [
          { blockType: "text", content },
          ...apiBlockRows(event.payload.block === undefined ? undefined : [event.payload.block]),
        ],
        tokenEstimate: textRaw(estimator, event.payload.text) + signatureTokens,
        signatureTokenEstimate: signatureTokens,
      };
    }
    case "model_change":
      return {
        blocks: [
          {
            blockType: "model_change",
            content: {
              previousModel: event.payload.previousModel,
              newModel: event.payload.newModel,
            },
          },
        ],
        tokenEstimate: textRaw(estimator, `${event.payload.previousModel} ${event.payload.newModel}`),
        signatureTokenEstimate: 0,
      };
    case "thinking_level_change":
      return {
        blocks: [
          {
            blockType: "thinking_level_change",
            content: {
              previousLevel: event.payload.previousLevel,
              newLevel: event.payload.newLevel,
            },
          },
        ],
        tokenEstimate: textRaw(estimator, `${event.payload.previousLevel} ${event.payload.newLevel}`),
        signatureTokenEstimate: 0,
      };
    case "tool_call":
      return {
        blocks: [
          {
            blockType: "tool_call",
            content: {
              toolCallId: event.payload.toolCallId,
              toolName: event.payload.toolName,
              arguments: event.payload.arguments,
            },
          },
          ...apiBlockRows(event.payload.block === undefined ? undefined : [event.payload.block]),
        ],
        // Tool calls count their serialized arguments.
        tokenEstimate: textRaw(estimator, JSON.stringify(event.payload.arguments)),
        signatureTokenEstimate: 0,
      };
    case "tool_result": {
      const content = textShaped(event.payload.content, event.payload.blocks);
      return {
        blocks: [
          {
            blockType: "tool_result",
            content: {
              toolCallId: event.payload.toolCallId,
              content,
              isError: event.payload.isError ?? false,
            },
          },
          ...apiBlockRows(event.payload.blocks),
        ],
        // Tool results count the full text-shaped content — the same string
        // block 0 carries in full — plus the blob estimate of any nested
        // image or document the text cannot see.
        tokenEstimate: textRaw(estimator, content) + blobTokens(event.payload.blocks),
        signatureTokenEstimate: 0,
      };
    }
    case "compact_continuation_marker": {
      // Typed marker: model-visible when served; not ordinary user chat.
      // Token estimate covers the stable model-facing instruction text.
      const content = {
        kind: event.payload.kind,
        continuationTurnId: event.payload.continuationTurnId,
        cause: event.payload.cause,
        action: event.payload.action,
        newUserRequest: event.payload.newUserRequest,
        waitForUser: event.payload.waitForUser,
      };
      const modelFacing = [
        "[compact continuation]",
        `cause=${event.payload.cause}`,
        `action=${event.payload.action}`,
        "newUserRequest=false",
        "waitForUser=false",
        `continuationTurnId=${event.payload.continuationTurnId}`,
      ].join(" ");
      return {
        blocks: [{ blockType: "compact_continuation_marker", content }],
        tokenEstimate: textRaw(estimator, modelFacing),
        signatureTokenEstimate: 0,
      };
    }
    case "turn_end":
      return null;
  }
}

File: packages/lhc-py/src/lhc/messages/internal/project.py
"""Ported from packages/lhc/src/messages/internal/project.ts.

Event to message + typed blocks. Verbatim means payload fields are copied
into block content untouched: nothing here trims, normalizes, splits, or
summarizes.
"""

from __future__ import annotations

import json
from dataclasses import dataclass

from ...shared_tech._jsstr import js_json_dumps
from ...shared_tech.token_counting import estimate_tokens

# Runtime import (not TYPE_CHECKING): Phase 2 bodies construct these records.
# Safe because the parent __init__ defines them BEFORE importing this module
# (see the IMPORT-ORDER CONSTRAINT note there).
from .. import Block, RecordedEvent


@dataclass(frozen=True, slots=True)
class ProjectedMessage:
    blocks: list[Block]
    token_estimate: int


def _copy_optional_string(
    content: dict[str, object], payload: object, key: str
) -> None:
    """Verbatim optional string: present only when the payload key is set."""
    if not isinstance(payload, dict) or key not in payload:
        return
    value = payload[key]
    if value is not None:
        content[key] = value


# turn_end is recorded in the event order but produces no message.
def project_event(event: RecordedEvent) -> ProjectedMessage | None:
    kind = event["eventKind"]
    payload = event["payload"]
    if kind in ("user_prompt", "runtime_note"):
        text = payload["text"]
        return ProjectedMessage(
            blocks=[Block(block_type="text", content={"text": text})],
            token_estimate=estimate_tokens(text),
        )
    if kind == "assistant_text":
        # providerUsage rides the event payload and is stored as a message
        # column (not a block) — projection leaves text + optional provenance.
        text = payload["text"]
        content: dict[str, object] = {"text": text}
        _copy_optional_string(content, payload, "provider")
        _copy_optional_string(content, payload, "model")
        _copy_optional_string(content, payload, "api")
        return ProjectedMessage(
            blocks=[Block(block_type="text", content=content)],
            token_estimate=estimate_tokens(text),
        )
    if kind == "assistant_thinking":
        # Verbatim payload copy: text always; signature + model identity when sent.
        text = payload["text"]
        content = {"text": text}
        _copy_optional_string(content, payload, "signature")
        _copy_optional_string(content, payload, "provider")
        _copy_optional_string(content, payload, "model")
        _copy_optional_string(content, payload, "api")
        # Count signature bytes too — when served back to the provider they sit in
        # the live context window (the fable live-vs-LHC token gap).
        signature = content.get("signature")
        if isinstance(signature, str) and signature != "":
            estimate_source = f"{text}{signature}"
        else:
            estimate_source = text
        return ProjectedMessage(
            blocks=[Block(block_type="text", content=content)],
            token_estimate=estimate_tokens(estimate_source),
        )
    if kind == "model_change":
        previous_model = payload["previousModel"]
        new_model = payload["newModel"]
        return ProjectedMessage(
            blocks=[
                Block(
                    block_type="model_change",
                    content={
                        "previousModel": previous_model,
                        "newModel": new_model,
                    },
                )
            ],
            token_estimate=estimate_tokens(f"{previous_model} {new_model}"),
        )
    if kind == "thinking_level_change":
        previous_level = payload["previousLevel"]
        new_level = payload["newLevel"]
        return ProjectedMessage(
            blocks=[
                Block(
                    block_type="thinking_level_change",
                    content={
                        "previousLevel": previous_level,
                        "newLevel": new_level,
                    },
                )
            ],
            token_estimate=estimate_tokens(f"{previous_level} {new_level}"),
        )
    if kind == "tool_call":
        arguments = payload["arguments"]
        serialized_arguments = js_json_dumps(arguments)
        return ProjectedMessage(
            blocks=[
                Block(
                    block_type="tool_call",
                    content={
                        "toolCallId": payload["toolCallId"],
                        "toolName": payload["toolName"],
                        "arguments": arguments,
                    },
                )
            ],
            token_estimate=estimate_tokens(serialized_arguments),
        )
    if kind == "tool_result":
        content = payload["content"]
        return ProjectedMessage(
            blocks=[
                Block(
                    block_type="tool_result",
                    content={
                        "toolCallId": payload["toolCallId"],
                        "content": content,
                        "isError": (
                            payload["isError"]
                            if payload.get("isError") is not None
                            else False
                        ),
                    },
                )
            ],
            token_estimate=estimate_tokens(content),
        )
    if kind == "turn_end":
        return None
    raise ValueError(f"unknown event kind {kind}")
